packages/js/components/src/experimental-select-tree-control/select-tree.tsx

```diff
@@ -0,0 +1,221 @@
+	items,
+	getItemLabel = ( item ) => item?.name || '',
+	getItemValue = ( item ) => item?.id || '',
+	getSelectedItemProps,
+	getFilteredItems,
+	treeRef: ref,
+	suffix = <SuffixIcon icon={ search } />,
+	placeholder,
+	isLoading,
+	onInputChange,
+	...props
+}: SelectTreeProps ) {
+	const filteredItems = getFilteredItems(
+		items,
+		props.createValue || '',
+		props.selected || [],
+		getItemLabel
+	);
+	const linkedTree = useLinkedTree( filteredItems );
+
+	const [ isFocused, setIsFocused ] = useState( false );
+
+	const comboBoxRef = useRef< HTMLDivElement >( null );
+
+	// getting the parent's parent div width to set the width of the dropdown
+	const comboBoxWidth =
+		comboBoxRef.current?.parentElement?.parentElement?.getBoundingClientRect()
+			.width;
+
+	const shouldItemBeExpanded = ( item: LinkedTree ): boolean => {
+		if ( ! props.createValue || ! item.children?.length ) return false;
+		return item.children.some( ( child ) => {
+			if (
+				new RegExp( props.createValue || '', 'ig' ).test(
+					child.data.name
+				)
+			) {
+				return true;
+			}
+			return shouldItemBeExpanded( child );
+		} );
+	};
+
+	return (
+		<Dropdown
+			className="woocommerce-experimental-select-tree-control__dropdown"
+			contentClassName="woocommerce-experimental-select-tree-control__dropdown-content"
+			focusOnMount={ false }
+			renderContent={ ( { onClose } ) =>
+				isLoading ? (
+					<div
+						style={ {
+							width: comboBoxWidth,
+						} }
+					>
+						<Spinner />
+					</div>
+				) : (
+					<Tree
+						{ ...props }
+						id={ `${ props.id }-menu` }
+						ref={ ref }
+						items={ linkedTree }
+						onTreeBlur={ onClose }
+						shouldItemBeExpanded={ shouldItemBeExpanded }
+						style={ {
+							width: comboBoxWidth,
+						} }
+					/>
+				)
+			}
+			renderToggle={ ( { isOpen, onToggle, onClose } ) => (
+				<div
+					className={ classNames(
+						'woocommerce-experimental-select-control',
+						{
+							'is-focused': isFocused,
+						}
+					) }
+				>
+					<label
+						htmlFor={ `${ props.id }-input` }
+						id={ `${ props.id }-label` }
+						className="woocommerce-experimental-select-control__label"
+					>
+						{ props.label }
+					</label>
+					<ComboBox
+						comboBoxProps={ {
+							className:
+								'woocommerce-experimental-select-control__combo-box-wrapper',
+							ref: comboBoxRef,
+							role: 'combobox',
+							'aria-expanded': isOpen,
+							'aria-haspopup': 'tree',
+							'aria-labelledby': `${ props.id }-label`,
+							'aria-owns': `${ props.id }-menu`,
+						} }
+						inputProps={ {
+							className:
+								'woocommerce-experimental-select-control__input',
+							id: `${ props.id }-input`,
+							'aria-autocomplete': 'list',
+							'aria-controls': `${ props.id }-menu`,
+							autoComplete: 'off',
+							onFocus: () => {
+								if ( ! isOpen ) {
+									onToggle();
+								}
+								setIsFocused( true );
+							},
+							onBlur: ( event ) => {
```

Comment:
I would move all these functionalities to a custom hook like `useCombobox` or something similar.